src/target/unix.rs

```diff
@@ -0,0 +1,42 @@
+use nix::{
+    sys::ptrace,
+    sys::wait::waitpid,
+    unistd::{execv, fork, ForkResult, Pid},
+};
+use std::ffi::CString;
+impl UnixTarget for Target {}
+pub struct Target {
+    pub pid: Pid,
+}
+
+pub trait UnixTarget {
+    /// Launches the debugee from the path provided
+    fn launch(path: &str) -> Result<Target, Box<dyn std::error::Error>> {
```

Comment:
This probably can return just `Result<Pid, Error>` instead of `Result<Target, ...>`, and then it can be moved to a `pub(crate)` free function (i.e. not necessarily be a part of the trait). Then, in `LinuxTarget`, we can use it as follows: 

```rust
struct LinuxTarget { pid: Pid }

impl LinuxTarget {
  fn launch(&self, path: &str) -> Result<LinuxTarget, ...> {
    let pid = target::unix::launch(path);
    LinuxTarget { pid }
  }
}
```

And then, `UnixTarget::launch` (the trait function) can just define a generic interface without a blanket implementation, e.g.:

```rust
trait UnixTarget {
  fn launch(&self, path: &str) -> Result<Box<dyn UnixTarget>, ...>;
```

What do you think?

Reply:
Thanks for the review. 😄 
I agree with the idea of making `launch` a free function, but then i couldn't quite grasp why we would need it in the `UnixTarget` too because the functionality of `launch` is wrapped in `LinuxTarget::launch`.
Am i missing something serious here, cause it feels like i am.
Could you please elaborate a bit more on this or guide me to some reading material ?

Reply:
>  Could you please elaborate a bit more on this

Sure! It boils down to the question of how do we create debug tools which can work on any supported target platform with a minimal number of changes required from a user. And you're right -- sorry for the confusion, `launch` should be a part of the concrete target type, actually. I was thinking in terms of using the `Target` trait as such:

```rust
trait Target {
  fn launch(path: &str) -> Box<dyn Target> where Self: Sized;
}

fn my_debug_tool<T: Target>() {
  let target = T::launch("/usr/bin/bash");
  target.read();
  // read target's memory etc.
}
```

but it doesn't make a lot of sense and it introduces unnecessary complexity. I think it's easier & better to instantiate the `Target` trait object before we call `my_debug_tool(target)`, i.e.:

```rust
fn my_debug_tool<T: Target>(target: T) {
  target.read();
  // read target's memory etc.
}

fn main() {
  let target = NativeTarget::launch("/usr/bin/bash");
  my_debug_tool(target);
}
```

So yeah, it's better to leave `fn launch(..)` out of traits altogether :+1:

Reply:
Thanks for the explanation.
I too feel that the second way is a bit more organic at least for me

```rust
fn my_debug_tool<T: Target>(target: T) {
  target.read();
  // read target's memory etc.
}

fn main() {
  let target = NativeTarget::launch("/usr/bin/bash");
  my_debug_tool(target);
}
```
So then i think this PR can be merged, if the review is done?

Regarding CIBadge for freeBSD we would have to do [little bit more steps](https://cirrus-ci.org/guide/quick-start/) and then update readme.
I can create a new issue for CirrusCi integration if required.

Reply:
Created #31 to track CirrusCI